### PR TITLE
Tutorial for testing a simple smart contract with Waffle

### PR DESCRIPTION
**Pull request by ekowalsk**
In this PR I add a new tutorial in developers/tutorials for testing a smart contract with Waffle

### CODE REVIEW

File: src/content/developers/tutorials/testing-ERC20-tokens-with-waffle/index.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ---
-title: Testing smart contracts with Waffle
+title: Testing ERC20 tokens with Waffle
```
Comment (samajammin) [inline src/content/developers/tutorials/testing-ERC20-tokens-with-waffle/index.md:2]:
```suggestion
title: Testing ERC-20 tokens with Waffle
```
Let's try to stick with a single convention: `ERC-20`

Comment (samajammin) [reply src/content/developers/tutorials/testing-ERC20-tokens-with-waffle/index.md:2]: Along with conventions, please keep the filename lowercase (i.e. `src/content/developers/tutorials/testing-erc-20-tokens-with-waffle/index.md`).

Comment (samajammin) [reply src/content/developers/tutorials/testing-ERC20-tokens-with-waffle/index.md:2]: Also please add `"erc-20"` as a tag to this tutorial 😄

**Comment by samajammin** [comment]
Thanks for the PR @ekowalsk!

@wackerow please review when able.

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+# Getting started
+The tutorial demonstrates test setup and run using yarn, but there is no problem if you prefer npm - I will provide proper references to the official Waffle [documentation](https://ethereum-waffle.readthedocs.io/en/latest/index.html). 
+## Install Waffle
+[Add](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#installation) ethereum-waffle package to the dev dependencies of your project.
+```bash
+yarn add --dev ethereum-waffle
+```
+## Example smart contract
+
+During the tutorial we'll work on a simple smart contract example - EtherSplitter. It does not much apart from allowing anyone to send some Wei and split it evenly between two predefined receivers.\
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:31]: Remove stray slash at the end

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+The split function requires the number of Wei to be even, otherwise it will revert. For both receivers it performs a Wei transfer followed by emission of the Transfer event.
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:32]: Been using convention of lowercase "wei" across the site (referencing the whitepaper)

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    event Transfer(address from, address to, uint256 amount);
+
+    constructor(address payable _address1, address payable _address2) public {
+        receiver1 = _address1;
+        receiver2 = _address2;
+    }
+    
+    function split() public payable {
+        require(msg.value % 2 == 0, 'Uneven Wei amount not allowed');
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:51]: Ditto

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  "compilerType": "solcjs",
+  "compilerVersion": "0.6.2",
+  "sourceDirectory": "./src",
+  "outputDirectory": "./build"
+}
+```
+Run ```yarn build```. As the result, the ``build`` directory will appear with the EtherSplitter compiled contract in JSON format.
+
+## Test setup
+Testing with Waffle requires using Chai matchers and Mocha, so you need to [add](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#writing-tests) them to your project. Update your package.json file and add the ``"test"`` entry in the scripts part:
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:79]: Single set of back-ticks should do around `"test"`

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+```json
+"scripts": {
+    "build": "waffle",
+    "test": "export NODE_ENV=test && mocha 'test/**/*.test.ts'"
+  }
+```
+If you want to [execute](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#running-tests) your tests, just run ```yarn test``` .
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:86]: Can use single back-ticks for the inline code snippet

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+# Testing
+Now create the ``test`` directory and create the new file ``test\EtherSplitter.test.ts``.
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:88]: Ditto

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+Next, we declare a variable called 'splitter' - this is our mock EtherSplitter contract. It is created before each execution of a single test by the ``deployContract`` method. This method simulates deployment of a contract from the wallet passed as the first parameter (sender's wallet in our case). The second parameter is the ABI and the bytecode of the tested contract - we pass there the json file of the compiled EtherSplitter contract from the `build` directory. The third parameter is an array with the contract's constructor arguments, which in our case, are the two addresses of the receivers.
+
+## ``changeBalances``
+
+First, we will check if the split method actually changes the balances of receivers' wallets. If we split 50 Wei from senders account, we would expect the balances of both receivers to increase by 25 Wei. We will use Waffle's ``changeBalances`` matcher:
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:118]: Ditto on "wei" (throughout)

File: src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  beforeEach(async () => {
+    splitter = await deployContract(sender, EtherSplitter, [receiver1.address, receiver2.address]);
+  });
+
+  // add the tests here
+
+});
+```
+A few words before we start.\
+The ``MockProvider`` comes up with a mock version of the blockchain. It also delivers mock wallets that will serve us for testing EtherSplitter contract. We can get up to ten wallets by calling ``getWallets()`` method on the provider. In the example, we get three wallets - for the sender and for two receivers.
```
Comment (wackerow) [inline src/content/developers/tutorials/waffle-test-simple-smart-contract/index.md:112]: Ditto on single back-ticks for inline code snippets (throughout)

**Review by wackerow — CHANGES_REQUESTED** [review]
@ekowalsk Thanks a lot for the PR! This looks great... no major changes, just some small semantic adjustments if you wouldn't mind touching them up =)

**Comment by ekowalsk** [comment]
Thank you for your reviews, I've applied all of your suggestions :)

**Review by wackerow — APPROVED** [review]
@ekowalsk Looks great, thanks again for the PR!

**Comment by samajammin** [comment]
@all-contributors please add @ekowalsk for content.

**Pull request merged**